File: LeetCode_Solution/136. Single Number.py
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        res = 0
        for i in nums:
            res = res ^ i
        return res
 
 # 按位运算符是把数字看做二进制数来进行计算
 # ^ 按位异或，相同得0，不同得1，遍历数组最后剩下的值是只出现过一次的值

File: java_solution/136. Single Number.java


// XOR，Bit Manipulation
// 0 ^ N = N
// N ^ N = 0
// Time complexity : O(n). We only iterate through nums, so the time complexity is the number of elements in nums.
// Space complexity : O(1).

class Solution {
    public int singleNumber(int[] nums) {
        
        int result = 0;
        
        for (int i = 0; i < nums.length; i++) {
            result ^= nums[i];
        }
        
        return result;
    }
}
